subprojects/core/src/main/java/org/gradle/internal/service/scopes/VirtualFileSystemServices.java

```diff
@@ -185,96 +164,50 @@ FileHasher createCachingFileHasher(StringInterner stringInterner, CrossBuildFile
             return fileHasher;
         }
 
-        FileWatcherRegistryFactory createFileWatcherRegistryFactory() {
+        WatchingVirtualFileSystem createVirtualFileSystem(
+            AdditiveCacheLocations additiveCacheLocations,
+            FileHasher hasher,
+            FileSystem fileSystem,
+            Stat stat,
+            StringInterner stringInterner
+        ) {
+            DefaultVirtualFileSystem delegate = new DefaultVirtualFileSystem(
+                hasher,
+                stringInterner,
+                stat,
+                fileSystem.isCaseSensitive() ? CASE_SENSITIVE : CASE_INSENSITIVE,
+                DirectoryScanner.getDefaultExcludes()
+            );
             OperatingSystem operatingSystem = OperatingSystem.current();
+            if (!(operatingSystem.isLinux() || operatingSystem.isMacOsX() || operatingSystem.isWindows())) {
+                return new WatchingNotSupportedVirtualFileSystem(delegate);
+            }
```

Comment:
I think we can do this better if we avoid doing the checks on the separate OSs twice. Instead, I'd introduce a method to `determineWatcherRegistryFactory(OperatingSystem)` which can return `null` or `Optional.empty()` for non-supported.

Reply:
I extracted the method.